common/djangoapps/third_party_auth/saml.py

```diff
@@ -246,85 +246,128 @@ def missing_variables(self):
-        the SAP SuccessFactors OData API.
+        Obtain a SAML assertion from the SAP SuccessFactors BizX OAuth2 identity provider service using
+        information specified in the third party authentication configuration "Advanced Settings" section.
         """
         session = requests.Session()
-        assertion = session.post(
-            self.sapsf_idp_url,
-            data={
-                'client_id': self.odata_client_id,
-                'user_id': user_id,
-                'token_url': self.sapsf_token_url,
-                'private_key': self.sapsf_private_key,
-            },
-            timeout=self.timeout,
-        )
-        assertion.raise_for_status()
-        assertion = assertion.text
-        token = session.post(
-            self.sapsf_token_url,
-            data={
-                'client_id': self.odata_client_id,
-                'company_id': self.odata_company_id,
-                'grant_type': 'urn:ietf:params:oauth:grant-type:saml2-bearer',
-                'assertion': assertion,
-            },
-            timeout=self.timeout,
-        )
-        token.raise_for_status()
-        token = token.json()['access_token']
-        session.headers.update({'Authorization': 'Bearer {}'.format(token), 'Accept': 'application/json'})
+        transaction_data = {
+            'operation_name': "generate_bizx_oauth_api_saml_assertion",
+            'endpoint_url': self.sapsf_idp_url,
+            'token_url': self.sapsf_token_url,
+            'company_id': self.odata_company_id,
+            'client_id': self.odata_client_id,
+            'user_id': user_id,
+            'private_key': self.sapsf_private_key,
+        }
+        try:
+            assertion = session.post(
+                transaction_data.get('endpoint_url'),
+                data=transaction_data,
+                timeout=self.timeout,
+            )
+            assertion.raise_for_status()
+        except requests.RequestException as err:
+            self.log_bizx_api_exception(transaction_data, err)
+            raise err
+        return assertion.text
+
+    def generate_bizx_oauth_api_access_token(self, user_id):
+        """
+        Request a new access token from the SuccessFactors BizX OAuth2 identity provider service
+        using a valid SAML assertion (see generate_bizx_api_saml_assertion) and the infomration specified
+        in the third party authentication configuration "Advanced Settings" section.
+        """
+        session = requests.Session()
+        transaction_data = {
+            'operation_name': "generate_bizx_oauth_api_access_token",
+            'endpoint_url': self.sapsf_token_url,
+            'client_id': self.odata_client_id,
+            'company_id': self.odata_company_id,
+            'grant_type': 'urn:ietf:params:oauth:grant-type:saml2-bearer',
+        }
+        try:
+            assertion = self.generate_bizx_oauth_api_saml_assertion(user_id)
```

Comment:
Move this line and line below outside of try/catch block as the method `generate_bizx_oauth_api_saml_assertion` has its own try/catch for `RequestException`.

Reply:
I had that originally, but since the function re-raises the exception I thought it should be moved inside of the try/catch -- thoughts?

Reply:
Yes, I agree that we should move these lines outside the try/catch here which will prevent us from double-logging.

Reply:
I guess I'm still getting caught up on the fact that we are re-raising the exception after logging it -- will this cause issues with the ultimate [return of the basic user details](https://github.com/edx/edx-platform/pull/16475/files#diff-31c7777edd76b5f6c6190122337c991cR370) in the case of an error?  We don't want to halt the workflow in this case, we just want to exit gracefully.

Reply:
Maybe I should rework the exception handling?

Reply:
Yeah, sounds like you don't want to raise within that logging helper function.